flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/async/Emitter.java

```diff
@@ -97,22 +102,18 @@ public void run() {
 		}
 	}
 
-	private void output(AsyncResult asyncResult) throws InterruptedException {
+	/**
+	 * Executed as a mail in the mailbox thread. Output needs to be guarded with checkpoint lock (for the time being).
+	 *
+	 * @param asyncResult the result to output.
+	 */
+	private void output(AsyncResult asyncResult) {
 		if (asyncResult.isWatermark()) {
-			synchronized (checkpointLock) {
-				AsyncWatermarkResult asyncWatermarkResult = asyncResult.asWatermark();
+			AsyncWatermarkResult asyncWatermarkResult = asyncResult.asWatermark();
```

Comment:
This action can be executed by a normal mailbox loop (not the one running in `AsyncWaitOperator` via `yieldToDownstream()`), that doesn't run letters under checkpoint lock (yet).

Also, I think you are losing guarantee, that this operation and `streamElementQueue.poll()` happen atomically.
It may happen, that task checkpoint, then restart and restore would emit duplicate events.

Reply:
Moved poll inside checkpoint lock